lib/shared/sdk/usbboot/protocol.js

```diff
@@ -0,0 +1,314 @@
+  // since the connection to the device is suddenly lost,
+  // but the module keeps waiting for it
+  return exports.sendBufferSize(device, NULL_BUFFER_SIZE)
+}
+
+/**
+ * @summary Read a buffer from a device
+ * @function
+ * @private
+ *
+ * @param {Object} device - device
+ * @param {Number} bytesToRead - bytes to read
+ * @fulfil {Buffer} - data
+ * @returns {Promise}
+ *
+ * @example
+ * usbboot.scan().then((devices) => {
+ *   return protocol.read(devices[0], 4).then((data) => {
+ *     console.log(data.readInt32BE())
+ *   })
+ * })
+ */
+exports.read = (device, bytesToRead) => {
+  return usb.performControlTransfer(device, {
+    /* eslint-disable no-bitwise */
+    bmRequestType: usb.LIBUSB_REQUEST_TYPE_VENDOR | usb.LIBUSB_ENDPOINT_IN,
+    wValue: bytesToRead & USBBOOT_MESSAGE_MAX_BUFFER_LENGTH,
+    wIndex: bytesToRead >> 16,
+    /* eslint-enable no-bitwise */
+
+    bRequest: USB_REQUEST_CODE_GET_STATUS,
+    length: bytesToRead
+  })
+}
+
+/**
+ * @summary Create a boot message buffer
+ * @function
+ * @private
+ *
+ * @description
+ * This is based on the following data structure:
+ *
+ * typedef struct MESSAGE_S {
+ *   int length;
+ *   unsigned char signature[20];
+ * } boot_message_t;
+ *
+ * This needs to be sent to the out endpoint of the USB device
+ * as a 24 bytes big-endian buffer where:
+ *
+ * - The first 4 bytes contain the size of the bootcode.bin buffer
+ * - The remaining 20 bytes contain the boot signature, which
+ *   we don't make use of in this implementation
+ *
+ * @param {Buffer} bootCodeBufferLength - bootcode.bin buffer length
+ * @returns {Buffer} boot message buffer
+ *
+ * @example
+ * const bootMessageBuffer = protocol.createBootMessageBuffer(50216)
+ */
+exports.createBootMessageBuffer = (bootCodeBufferLength) => {
+  const bootMessageBufferSize = BOOT_MESSAGE_BOOTCODE_LENGTH_SIZE + BOOT_MESSAGE_SIGNATURE_SIZE
+
+  // Buffers are automatically filled with zero bytes
+  const bootMessageBuffer = Buffer.alloc(bootMessageBufferSize)
+
+  // The bootcode length should be stored in 4 big-endian bytes
+  bootMessageBuffer.writeInt32BE(bootCodeBufferLength, BOOT_MESSAGE_BOOTCODE_LENGTH_OFFSET)
+
+  return bootMessageBuffer
+}
+
+exports.parseFileMessageBuffer = (fileMessageBuffer) => {
+  const commandCode = fileMessageBuffer.readInt32LE(FILE_MESSAGE_COMMAND_OFFSET)
+  const command = _.nth([
+    exports.FILE_MESSAGE_COMMAND_GET_FILE_SIZE,
+    exports.FILE_MESSAGE_COMMAND_READ_FILE,
+    exports.FILE_MESSAGE_COMMAND_DONE
+  ], commandCode)
+
+  if (_.isNil(command)) {
+    throw new Error(`Invalid file message command code: ${commandCode}`)
+  }
+
+  const fileName = _.chain(fileMessageBuffer.toString('ascii', FILE_MESSAGE_FILE_NAME_OFFSET))
+
+    // The parsed string will likely contain tons of trailing
+    // null bytes that we should get rid of for convenience
+    // See https://github.com/nodejs/node/issues/4775
+    .takeWhile((character) => {
+      return character !== '\0'
+    })
```

Comment:
> Alternatively, to more closely mimic the upstream C implementation, I guess we could assume that the first null byte is the end-of-string marker and simply discard everything after that?

That's exactly what its doing. The `takeWhile` function will consume all the characters it can before hitting the first null byte, and discard everything else.